src/apps/replication/meta_server/server_state.cpp

```diff
@@ -398,73 +547,34 @@ error_code server_state::initialize_apps()
         return err;
     }
 
-    err = ERR_OK;
-    // create cluster_root/apps/app_name node
-    std::string app_path = get_app_path(app);
-    blob value;
-    marshall_json(value, app, true);
-
-    _storage->create_node(app_path, LPC_META_STATE_SVC_CALLBACK,
-        [&err, this, &app, &tracker, &app_path](error_code ec)
-        {
-            if (ec == ERR_OK)
-            {
-                int32_t max_replica_count = (int)dsn_config_get_value_uint64("replication.app",
-                    "max_replica_count", 3, "maximum replica count for each partition");
-
-                for (int i = 0; i < app.partition_count; i++)
-                {
-                    std::string partition_path = join_path(app_path, boost::lexical_cast<std::string>(i));
-
-                    partition_configuration ps;
-                    ps.app_type = app.app_type;
-                    ps.ballot = 0;
-                    ps.gpid.app_id = app.app_id;
-                    ps.gpid.pidx = i;
-                    ps.last_committed_decree = 0;
-                    ps.max_replica_count = max_replica_count;
-                    ps.primary.set_invalid();
-
-                    app.partitions.push_back(ps);
-
-                    blob value;
-                    marshall_json(value, ps);
-
-                    _storage->create_node(partition_path, LPC_META_STATE_SVC_CALLBACK,
-                        [&err, this, &app](error_code ec)
-                        {
-                            if (ec == ERR_OK)
-                            {
-
-                            }
-                            else
-                            {
-                                err = ec;
-                            }
-                        },
-                        value,
-                        &tracker
-                        );
-                }
-            }
-            else
-            {
-                err = ec;
-            }
-        },
-        value,
-        &tracker
-        );
-    
-    // wait for all those tasks completed
-    dsn_task_tracker_wait_all(tracker.tracker());
+    typedef dsn::dist::meta_state_service::transaction_entries TEntries;
+    dsn::blob data;
 
-    if (err == ERR_OK)
+    for (const app_state& app: _apps)
     {
-        _apps.push_back(app);
+        std::shared_ptr<TEntries> entries = _storage->new_transaction_entries(app.partition_count+1);
+        marshall_json(data, app, app.status==app_status::available);
+        entries->create_node(get_app_path(app), data);
+        for (const partition_configuration& pc: app.partitions) {
+            marshall_json(data, pc);
+            entries->create_node(get_partition_path(pc.gpid), data);
+        }
+        error_code ec;
+        task_ptr tsk = _storage->submit_transaction(
```

Comment:
Is there upper bound about the size of the transaction? I'm worried that the partition count can be very large in real settings, e.g., tens of thousands.